src/lookup/lookup.ts

```diff
@@ -0,0 +1,118 @@
+import {
+  Catalog,
+  Dataset,
+  Distribution,
+  IRI,
+} from '@netwerk-digitaal-erfgoed/network-of-terms-catalog';
+import {
+  Error,
+  QueryTermsService,
+  ServerError,
+  Terms,
+  TermsResult,
+  TimeoutError,
+} from '../services/query';
+import {Term} from '../services/terms';
+
+export type LookupQueryResult = {
+  uri: IRI;
+  distribution: SourceResult;
+  result: LookupResult;
+};
+
+export type SourceResult = Distribution | SourceNotFoundError;
+
+export type LookupResult = Term | NotFoundError | TimeoutError | ServerError;
+
+export class SourceNotFoundError {
+  readonly message: string;
+
+  constructor(readonly iri: IRI) {
+    this.message = `No source found that can provide term with URI ${iri}`;
+  }
+}
+
+export class NotFoundError {
+  readonly message: string;
+
+  constructor(readonly iri: IRI) {
+    this.message = `No term found with URI ${iri}`;
+  }
+}
+
+export class LookupService {
+  constructor(
+    private catalog: Catalog,
+    private queryService: QueryTermsService
+  ) {}
+
+  public async lookup(
+    iris: IRI[],
+    timeoutMs: number
+  ): Promise<LookupQueryResult[]> {
+    const irisToDataset = iris.reduce((acc, iri) => {
+      const dataset = this.catalog.getDatasetByTermIri(iri);
+      if (dataset) {
+        acc.set(iri, dataset);
+      }
+      return acc;
+    }, new Map<IRI, Dataset>());
+
+    const datasetToIris = [...irisToDataset].reduce(
+      (datasetMap, [iri, dataset]) => {
+        datasetMap.set(dataset, [...(datasetMap.get(dataset) ?? []), iri]);
+        return datasetMap;
+      },
+      new Map<Dataset, IRI[]>()
+    );
+
+    const lookups = <Promise<TermsResult>[]>[];
+    datasetToIris.forEach((iris, dataset) =>
```

Comment:
Or can we use a `map` here? For example: `const lookups = datasetToIris.map(...)`

Reply:
Yes, that’s much cleaner. FYI, applying `.map` to a Map isn’t possible; you have to spread into an array first – which I overlooked and so used a `.forEach` where none was needed:

```javascript
[...datasetToIris].map(...)
```